contractsV5/eth2dai/Eth2DaiReserve.sol

```diff
@@ -0,0 +1,871 @@
+        delete tokenBasicData[tokenAddr];
+
+        emit TokenDelisted(token);
+    }
+
+    event FeeBpsSet(uint feeBps);
+
+    function setFeeBps(uint _feeBps) public onlyAdmin {
+        require(_feeBps < 10000, "setFeeBps: feeBps >= 10000");
+
+        feeBps = _feeBps;
+        emit FeeBpsSet(feeBps);
+    }
+
+    function showBestOffers(ERC20 token, bool isEthToToken, uint srcAmountToken) public view
+        returns(uint destAmount, uint destAmountToken, uint [] memory offerIds) 
+    {
+        OfferData [] memory offers;
+        ERC20 dstToken = isEthToToken ? token : wethToken;
+        ERC20 srcToken = isEthToToken ? wethToken : token;
+
+        OfferData memory bid;
+        OfferData memory ask;
+        (bid, ask) = getFirstBidAndAskOrders(token);
+
+        (destAmount, offers) = findBestOffers(dstToken, srcToken, (srcAmountToken * 10 ** 18), bid, ask);
+        
+        destAmountToken = destAmount / 10 ** 18;
+        
+        uint i;
+        for (i; i < offers.length; i++) {
+            if (offers[i].id == 0) {
+                break;
+            }
+        }
+    
+        offerIds = new uint[](i);
+        for (i = 0; i < offerIds.length; i++) {
+            offerIds[i] = offers[i].id;
+        }
+    }    
+    
+    function findBestOffers(ERC20 dstToken, ERC20 srcToken, uint srcAmount, OfferData memory bid, OfferData memory ask)
+        internal view
+        returns(uint totalDestAmount, OfferData [] memory offers)
+    {
+        uint remainingSrcAmount = srcAmount;
+        uint maxOrdersToTake;
+        uint maxTraversedOrders;
+        uint minPayAmount;
+        uint numTakenOffer = 0;
+        totalDestAmount = 0;
+        ERC20 token = srcToken == wethToken ? dstToken : srcToken;
+
+        (maxOrdersToTake, maxTraversedOrders, minPayAmount) = calcOfferLimitsFromFactorData(
+            token,
+            (srcToken == wethToken),
+            bid,
+            ask,
+            srcAmount
+        );
+
+        offers = new OfferData[](maxTraversedOrders);
+
+        // return earlier, we don't want to take any orders
+        if (maxTraversedOrders == 0 || maxOrdersToTake == 0) {
+            return (totalDestAmount, offers);
+        }
+
+        // otc's terminology is of offer maker, so their sellGem is our (the taker's) dest token.
+        // if we don't have first offer, try to get it
+        if ((srcToken == wethToken && bid.id == 0) || (dstToken == wethToken && ask.id == 0)) {
+            offers[0].id = otc.getBestOffer(dstToken, srcToken);
+            // assuming pay amount is taker pay amount. (in otc it is used differently)
+            (offers[0].buyAmount, , offers[0].payAmount, ) = otc.getOffer(offers[0].id);
+        } else {
+            offers[0] = srcToken == wethToken ? bid : ask;
+        }
+
+        // putting here so if src amount is 0, we won't revert and still consider the first order as rate
+        if (remainingSrcAmount == 0) { return (totalDestAmount, offers); }
```

Comment:
I think this check can be added to line 515 checks.

Reply:
I put here to make sure `offers` has at least 1 offer, `totalDestAmount` is 0 but offers has an offer, so we could know about the rate.

Reply:
OK
how do we use this rate?
it will not return 0 rate according to zero totalDestAmount

Reply:
Well, for now it is only for `showBestOffers` function to test, our `getConversionRate` func already avoid calling with `srcQty = 0`. If we remove `showBestOffers`, then we could safely remove this check as well.

Reply:
so lets not leave stuff for debugging.
see in show best offers if can always call with srcQty > 0?